src/ring_theory/eisenstein_criterion.lean

```diff
@@ -0,0 +1,100 @@
+/-
+Copyright (c) 2020 Chris Hughes. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Chris Hughes
+-/
+import ring_theory.ideal_operations
+import data.polynomial
+import tactic.apply_fun
+import ring_theory.prime
+/-!
+# Eisenstein's criterion
+
+A proof of a slight generalisation of Eisenstein's criterion for the irreducibility of
+a polynomial over an integral domain.
+-/
+
+open polynomial ideal.quotient
+variables {R : Type*} [integral_domain R]
+
+/-- If `f` is a non constant polynomial with coefficients in `R`, and `P` is a prime ideal in `R`,
+then if every coefficient in `R` except the leading coefficient is in `P`, and
+the trailing coefficient is not in `P^2` and no non units in `R` divide `f`, then `f` is
+irreducible. -/
+theorem irreducible_of_eisenstein_criterion {f : polynomial R} {P : ideal R} (hP : P.is_prime)
+  (hfl : f.leading_coeff ∉ P)
+  (hfP : ∀ n : ℕ, ↑n < degree f → f.coeff n ∈ P)
+  (hfd0 : 0 < degree f) (h0 : f.coeff 0 ∉ P^2)
+  (hu : ∀ x : R, C x ∣ f → is_unit x) : irreducible f :=
+have hf0 : f ≠ 0, from λ _, by simp * at *,
```

Comment:
Are there no reusable sublemmas in this 70-line proof?

Reply:
Not really. A lot of what's there is there twice because I have to prove the same things about `p` and `q`

Reply:
Well... that sounds like you could extract it into an `_aux` or two. It would clarify the overall structure, I think.